Refactor BuilderBase to account for changes to shell, shebang key in script schema

shahzebsiddiqui opened the pull request:
sync script schema from https://github.com/buildtesters/schemas/blob/master/script/script-v0.0.1.schema.json
modify implementation on shell detection, environment variable, and the
command to execute the test. We can now support passing shell options on
the command line. The shebang can be customized via 'shebang' key if
it's not found it will use the default shebang based on shell type. If
it's python it will resort to path of 'python' wrapper.
When running in dry mode we were getting output header of table of
results that were not suppose to show up.
Add more log details for BuilderBase

shahzebsiddiqui committed:
sync script schema from
https://github.com/buildtesters/schemas/blob/master/script/script-v0.0.1.schema.json
modify implementation on shell detection, environment variable, and the
command to execute the test. We can now support passing shell options on
the command line. The shebang can be customized via 'shebang' key if
it's not found it will use the default shebang based on shell type. If
it's python it will resort to path of 'python' wrapper.
When running in dry mode we were getting output header of table of
results that were not suppose to show up.
Add more log details for BuilderBase

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -266,6 +267,36 @@ def __init__(self, name, recipe, buildspec=None, testdir=None):
 
+        self.shell = self.recipe.get("shell")
+        # if 'shell' key not found in Buildspec, let's set it to BUILDTEST_SHELL and log a message
+        if not self.shell:
```
Do you want to make self.shell a property instead? I can see this being reused.

shahzebsiddiqui replied:
Can you describe with example how you plan to reuse the shell decorator. I have been getting up to speed on decorators but haven't found a use-case. 

I am thinking you want to invoke BuilderBase class directly if i understand. I think changing shell from ``bash`` --> ``sh`` or ``sh`` --> ``bash`` would be okay for most part, but there are some minor differences between the two shells that can cause an error depending on how ``run`` section is setup. I was able to find this link https://www.gnu.org/software/bash/manual/html_node/Major-Differences-From-The-Bourne-Shell.html outlining some of the differences between ``bash`` and ``sh`` 

I don't see a usecase where we can allow a shell to change from ``bash`` or ``sh`` --> ``python`` since that would cause run section to break pretty dramatically.

vsoch replied:
All the times in the code we do some action like "self.recipe.get("shell") we would want to just reference self.shell. And I was thinking this morning, since there are now a bunch of different parameters around a shell, would it make sense to have it be a class? E.g.,

```Python
self.shell = Shell(self.recipe.get('shell', '/bin/bash'))
self.shell.opts
self.shell.shebang
```
etc.

shahzebsiddiqui replied:
I have been thinking about the ``shell`` key this is specific to ``LocalExecutor`` behavior since we are running test locally. This wont be relevant when running test with ``SlurmExector``. It's true we still need to detect the shell for writing the test, get test extension,  but execution of test is depended on the executor type. We also have ``executor`` key https://github.com/buildtesters/buildtest/blob/devel/buildtest/buildsystem/schemas/script/script-v0.0.1.schema.json#L39  which we have not implemented during the build stage. Right now everything defaults to LocalExecutor but again this is out of scope for this PR but worth pointing out.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -266,6 +267,36 @@ def __init__(self, name, recipe, buildspec=None, testdir=None):
+        if not self.shebang:
+            # when shell: python we set shebang to location of 'python' wrapper
+            if "python" in self.shell_name:
+                self.shebang = "#!" + shutil.which("python")
```
This will trigger an error if Python isn’t found and returned as None.

vsoch replied:
Wouldn’t there be a more elegant way to do this instead of hard coding python?

shahzebsiddiqui replied:
Yea if python is not found we can raise an error. Also ``shell: python`` adapts to both python2 or python3 but it's not clear to user when they write in ``run`` section and test is not portable since it depends on user python environment. So if it works for me it may not necessarily work for you.

Right now we don't have ability to pass system python let's say ``shell: /usr/bin/python`` or ``shell: /usr/bin/python3`` similarly.

I particularly don't like having ``env`` support when shell is python, it causes some confusion to user when it sees buildtest modify the python script. I was thinking of removing this support but didnt for time being

shahzebsiddiqui replied:
I don't think it's possible for ``python`` to not be found when running buildtest. It would cause an error running buildtest. In the ``master`` branch we had the **buildtest** https://github.com/buildtesters/buildtest/blob/master/bin/buildtest as a shell script that checks for the python client as a requirement. That has been removed in ``devel`` but i can see it being useful having these checks at start of buildtest.

vsoch replied:
We only have python3 on sherlock, which is what made me think of this. I think (given that there can be multiple shells allowed) instead of hard coding them in specifically, it should be done with a single list, the schema validates, but then the code extracts (whatever the valid thing is). That's a much more elegant implementation than hard coding python into the code (and all the other future cases).

shahzebsiddiqui replied:
Are you referring to modifying the schema line https://github.com/buildtesters/schemas/blob/master/script/script-v0.0.1.schema.json#L44 

Do you want add python3 and system python (/bin/python, /bin/python3) into the pattern

```
"pattern": "^(/bin/bash|/bin/sh|sh|bash|python|python3|/bin/python|/bin/python3).*"
```

The python interpreter will be whatever python wrapper was used when user is invoking buildtest so it's not possible to not have python in $PATH. I think having the distinction between python3 and python2 would be useful so user knows how to write their script in ``run`` section.

vsoch replied:
I think I've mentioned this before - in that shell is an executable, I don't think we should limit the user to some known set. I should be able to run bash, sh (or a specific location, /bin/sh) or python3, or python4 when it comes out. We shouldn't need to pre-define the executables that the use can specify, only look for them with shutil.which() and exit with a clear error if they chose something not findable.

shahzebsiddiqui replied:
If i understand you like to run ``shutil.which()`` on ``self.shell_name`` and if its ``None`` raise an error.  This would be much earlier in the code, but note that we have ``BuildTestCommand`` raise this error at https://github.com/buildtesters/buildtest/blob/devel/buildtest/utils/command.py#L46 which would eventually catch this exception.

vsoch replied:
Yep.

shahzebsiddiqui replied:
See https://github.com/buildtesters/buildtest/pull/290/commits/922756693cf0320010121c16e3deb243b2c756ba

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -319,8 +351,7 @@ def get_environment(self):
         pairs = self.recipe.get("env")
-        shell = self.get_shell()
-
+        shell = self.shell_name
```
This would use the self.shell property.

shahzebsiddiqui replied:
I dont feel comfortable making this change, i think this relates to discussion  https://github.com/buildtesters/buildtest/pull/290#discussion_r417504156

vsoch replied:
Yep happy to review this again in another PR.

shahzebsiddiqui committed:
raise exception BuildTestError when self.shell_name (python, sh, bash)
not found in system. We need to terminate early since we can't run test
without it.

shahzebsiddiqui committed:
applying black

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

shahzebsiddiqui committed:
Add Shell class that returns shell program, options, shell path, and
shebang. This class is intended to get the shell type
based on 'shell' key in Buildspec. The Shell class is not integrated
with BuilderBase class at this point.

vsoch commented on buildtest/utils/shell.py:
```diff
@@ -0,0 +1,67 @@
+from buildtest.exceptions import BuildTestError
+
+
+class Shell:
```
Perhaps a default to bash?

shahzebsiddiqui replied:
i can do that so make argument ``shell`` as optional argument so that we can invoke the class as
`` shell = Shell()`` which defaults to bash.

shahzebsiddiqui replied:
So how is this

```
>>> shell = Shell()
>>> shell.get()
{'name': 'bash', 'opts': '', 'path': '/usr/bin/bash', 'shebang': '#!/usr/bin/bash'}
>>> sh = Shell("sh")
>>> sh.get()
{'name': 'sh', 'opts': '', 'path': '/usr/bin/sh', 'shebang': '#!/usr/bin/sh'}

```

vsoch commented on buildtest/utils/shell.py:
```diff
@@ -0,0 +1,67 @@
+
+        self.name = shell.split()[0]
+        self.opts = " ".join(shell.split()[1:])
+
```
The function opts would immediately override the class variable of the same name.

shahzebsiddiqui replied:
so what are you suggesting we do? Remove ``opts`` method?

vsoch replied:
It would be appropriate to parse original opts in the init, but then have this function more properly named based on what it does, e.g., shell.set_opts()

vsoch commented on buildtest/utils/shell.py:
```diff
@@ -0,0 +1,67 @@
+        path = shutil.which(self.name)
+
+        # raise an exception if shell program is not found
+        if not path:
```
These should be done in init.

shahzebsiddiqui replied:
ok i have done this in init also, but found that i also need this check in ``path`` method when there is a change in name attribute meaning ``shell.name = "python1"`` we report an error otherwise the path and shebang will be ``None``

vsoch replied:
Just create a function to handle these operations and call it in init. If the user of the Shell class wants to call it again after, they are free to do that.

vsoch commented on buildtest/utils/shell.py:
```diff
@@ -0,0 +1,67 @@
+        self.shebang = f"#!{path}"
+        return path
+
+    def get(self):
```
Test this out - most objects already can be serialized as dict so it’s about setting properties.

shahzebsiddiqui committed:
make "bash" as default shell so we can invoke Shell() without any
argument.
Raise exception in init method when path is not found. This raises an
error early as pose to later when invoking Shell.get()

vsoch committed:
Merge branch 'update_shell_logic' of github.com:shahzebsiddiqui/buildtest-framework into update_shell_logic

vsoch committed:
fixing up shell parser to use shared setter and functions

Signed-off-by: vsoch <vsochat@stanford.edu>

shahzebsiddiqui committed:
integrate Shell with BuilderBase class.
Remove env support detection when Buildspec has 'shell: python'. Should
make note that 'env' works for shell scripts and not python scripts.

shahzebsiddiqui committed:
applying black

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -265,6 +268,19 @@ def __init__(self, name, recipe, buildspec=None, testdir=None):
+        # if 'shell' key  is defined in Buildspec let's override the default shell
+        if self.recipe.get("shell"):
+            shell = Shell(self.recipe.get("shell"))
+
```
These 5 lines can just be

```python
shell = Shell(self.recipe.get("shell", "bash"))
```

shahzebsiddiqui replied:
This is now fixed along with the other stuff mentioned, feel free to resolve what you think is addressed.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -265,6 +268,19 @@ def __init__(self, name, recipe, buildspec=None, testdir=None):
+        self.shell = shell.get()
+
+        # set shebang to value defined in Buildspec, if not defined then get one from Shell class
+        self.shebang = self.recipe.get("shebang") or self.shell["shebang"]
```
The reason to have the self.shell as the shell class is that we would do:

```
self.shell = Shell()
```
and then refer to `self.shell.shebang` and not need to do shell.get() (this doesn't make sense to me). If the shebang can be included with the shell init and then referenced in the same way, this would also be cleaner code.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -319,11 +336,7 @@ def get_environment(self):
-        # If we are using a python shell, we need to import os first
-        if "python" in shell:
-            env.append("import os")
+        shell = self.shell["name"]
```
This would be `self.shell.name`

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -319,11 +336,7 @@ def get_environment(self):
 
         env = []
         pairs = self.recipe.get("env")
```
If you do self.recipe.get("env", []) you don't need to do `if pairs` but can just loop through them still (an empty list).

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -393,7 +397,7 @@ class data structures that are used by both run and dry_run
 
         # Get the shell (sh, bash) for writing testscript. A Buildspec could specify this via ``shell`` key
-        self.metadata["shell"] = self.get_shell()
+        self.metadata["shell"] = self.shell
```
Why is this added to metadata?

shahzebsiddiqui replied:
Ill remove it for now, it was not clear how this metadata will be used going forward one thought was to use the metadata to write to a json file for reporting (``buildtest build report``) #284

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -498,12 +502,14 @@ def run_tests(self, testfile):
+        # command to run the test
+        cmd = []
+        # build the run command that includes the shell path, shell options and path to test file
+        cmd += [self.shell["path"], self.shell["opts"], testfile]
```
This again should reference properties and not dictionary items, given that you are interacting with the class.

shahzebsiddiqui replied:
This is fixed now

vsoch commented on buildtest/utils/command.py:
```diff
@@ -54,27 +116,26 @@ def execute(self):
-                self.out.append(out)
-            if err:
-                self.err.append(out)
+        # Capturing provides temporary output and error files
```
Why are these changes showing up here when I already made them in another PR? I wrote these lines...

shahzebsiddiqui replied:
I was testing something, this was an accident my bad ill undo this, ill copy the file from devel since i didnt make any change to the file.

shahzebsiddiqui committed:
address fix reported by @vsoch

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -265,6 +268,19 @@ def __init__(self, name, recipe, buildspec=None, testdir=None):
                 % (self.type, self.recipe.get("type"))
             )
+        # The default shell will be bash
+        self.shell = Shell(self.recipe.get("shell", "bash"))
```
This whole section is much cleaner than it was before - good job!

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -318,36 +335,23 @@ def get_environment(self):
-            # Handles bash and sh
-            if re.search("(bash|sh)$", shell):
-                [env.append("%s=%s" % (k, v)) for k, v in pairs.items()]
+        # if pairs:
```
This is commented out and should not be there.

shahzebsiddiqui replied:
See https://github.com/buildtesters/buildtest/pull/290#commitcomment-38882276. I introduced this bug by commenting this out. I guess we still need the condition ``if pairs`` when a Buildspec doesn't define ``env`` section we just skip the whole thing

vsoch commented on buildtest/utils/command.py:
```diff
@@ -1,7 +1,69 @@
 import subprocess
 import shlex
 import shutil
+import tempfile
```
This was written by me, and shouldn't appear like this. It suggests that you didn't pull the merged commits from my contribution. It's not really okay to take credit for my work, so this needs to be fixed.

shahzebsiddiqui replied:
I am not sure what happened i do see your git commit **Fix/command out err (#289)** in my commit history https://github.com/shahzebsiddiqui/buildtest-framework/commits/update_shell_logic 
The only other explanation could be that when i copy-paste to fix this i applied black so it's showing a diff.
I am bit afraid to do a merge from devel now

shahzebsiddiqui committed:
Fix two bugs introduced in previous commit.
Need to add self.metadata["shell"] because it was referenced in
test_base.py
The second bug was when a Buildspec doesn't contain 'env' it causes
a bug during environment section, we needed to skip process any keys
since there is no items to process

shahzebsiddiqui committed:
Merge branch 'devel' into update_shell_logic

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -498,12 +500,14 @@ def run_tests(self, testfile):
-        cmd = [self.get_shell(), testfile]
+        # command to run the test
+        cmd = []
+        # build the run command that includes the shell path, shell options and path to test file
```
If the cmd is created here and then has lines added to it, why do we need to create it as an empty list first? E.g., just do:

```python
cmd = [self.shell.path, self.shell.opts, testfile]
```

shahzebsiddiqui replied:
This is fixed now in https://github.com/shahzebsiddiqui/buildtest-framework/commit/b18e1c53fbb42e081b6e016daa911d63677a9295

shahzebsiddiqui committed:
add regression test for Shell class, tested this locally, should
expect  100% coverage.
